### ISSUE TITLE
Feature Request: Pause playback before toggling "Fixed Volume 100%" Audio setting

### ISSUE THREAD

**Issue by SamInPgh** [kaaholst/android-squeezer#710]
I have recently implemented support in the Denon AVP/AVR plugin to dynamically detect user changes to this player setting and make adjustments on the fly with regard to how subsequent "mixer volume" commands are processed.  The problem is that there can be a lag of up to a second or two before the plugin is notified of the change via its registered LMS callback function and, during that time, the playback volume can, under certain conditions, spike drastically before the appropriate corrective actions can be taken. Although this situation will hopefully be rarely seen as users are cautioned not to change the setting while playing music, doing so could conceivably cause speaker damage. Having failed to discover a way to address this at the LMS player level, I am approaching client developers (you are the first) to make this request, as the change would be relatively simple there. Although the change would admittedly affect all users, I doubt that anyone would object to it as the potential problem it addresses is theoretically not limited to users of the Denon plugin.

**Comment by kaaholst** [comment]
I think I understand the request. It's, as you say "relatively" simple to implement in the client. 

It is however a bit of a hack. The reason is that Squeezer does not implement any of the actions performed from the players menus. We request the home menu for the player via a `["menu", "direct:1"]` command. We then display the returned menu items, and perform the actions in them, when an item is selected. This means that Squeezer does not actually know when the "Fixed Volume 100%" Audio setting is toggled. 

Having said that, I do agree that no one would probably object. It's visible what happens, and it's quite easy to start playback again.

I have noted that we receive an item as below for the "Fixed Volume 100%" toggle, so I can pause playback if cmd starts with: `["playerpref", "digitalVolumeControl"]`.
```
{
  "checkbox": 0,
  "text": "Fixed Volume 100%",
  "actions": {
    "off": {
      "cmd": [
        "playerpref",
        "digitalVolumeControl",
        1
      ],
      "player": 0
    },
    "on": {
      "cmd": [
        "playerpref",
        "digitalVolumeControl",
        0
      ],
      "player": 0
    }
  }
}
```

Would that, in your opinion, be an appropriate way to proceed?

**Comment by SamInPgh** [comment]
Yes, I believe that would be exactly the way to proceed, assuming that the client is able to take the pause action prior to the "playerpref digitalVolumeControl" command being executed by LMS. Is that the case? The plugin currently registers to receive all LMS "playerpref" updates but is not notified of the "digitalVolumeControl" change until after the command has been executed, resulting in the temporary volume spike.

**Comment by kaaholst** [comment]
Yes I send the pause command before the digitalVolumeControl command.

If you like, you can test using this [apk](https://drive.google.com/file/d/1xanOkuQpy36v2aw4OIm1brkCriPII6s_/view?usp=sharing)

**Comment by SamInPgh** [comment]
Well, I have good news and bad news. I see the pause event in my log,
followed shortly thereafter (~0.25 secs) by the "playerpref
digitalVolumeControl', so your change is working. However, I am still
experiencing a short volume spike before the audio is paused, which I am
now thinking must be due to some sort of buffering lag. I don't know how
much time you want to spend on this admittedly fringe scenario but if you
have an easy way of waiting for a short interval of time (0.5 secs?) after
the pause before sending the digitalVolumeControl command, it would at
least be interesting to see what effect it might have. If not, I completely
understand. Thanks.

**Comment by kaaholst** [comment]
For debugging purposes I've added a delay of 500ms after the pause command. Other delays I suspect comes from network or LMS.

**Comment by SamInPgh** [comment]
That works great! No volume spiking. As a side effect, the delay after
pausing also seems to cause the LMS player volume to be set to 0 when
toggling from fixed to variable output, which is not a bad thing as the
user will have to readjust it anyway. Must be something in the LMS player
code. Out of curiosity, I checked this using another client by pausing
playback before toggling to variable output and then resuming playback.
Sure enough, the player volume was also set to 0. It may also happen when
toggling to fixed 100% output but I don't see that behavior because the
plugin sets the volume to 25% when it detects toggling in that direction.

The bottom line is that I am happy with this change and hope that you can
incorporate it into the app. Thanks!

**Comment by kaaholst** [comment]
Great this it's now working. However, I do have concerns about the implementation. I generally avoid sleeping for arbitrary intervals, e.g. there are no sleeps in the Squeezer source code (up until now at least). 

In this case we know that a delay of 500ms is appropriate for your particular setup of hardware, software and infrastructure. We don't know if it's appropriate or even necessary in other environments.

I would prefer if we can find a better way to synchronize this.

**Comment by SamInPgh** [comment]
I understand your concerns, although my testing with another instance of
Squeezelite using Bluetooth output also worked well with the change. What
would you think about trying muting the playback instead of pausing,
without the intervening delay of course? Although it may have the same
buffering lag issue, it's possible that it might not, as the code path for
muting would be different. Just a thought. In the meantime, I will continue
to look for a place where this might be done in the LMS player code.

**Comment by kaaholst** [comment]
I've updated the link with a version which mutes instead of pause before the digitalVolumeControl command (with no delay) .

**Note** this is the same link, so copy the current apk before you download this, if you want to test both pause and muting..

It doesn't work as expected in my opinion though. It doesn't mute when it's set to fixed volume 100%. But when it's unchecked it does mute. I think these changes conflicts with some corrective actions the LMS is performing.

I think you should check the LMS code (or use the forum) so that our changes do not counteract the corrective actions that the LMS is attempting.

**Comment by SamInPgh** [comment]
The muting version still suffers from the buffering delay and resultant
volume spike that the pause does without the sleep. And yes, there are some
"twisty little passages" in the LMS muting logic that I'd rather not get
involved with unless there's no other option. So it seems that the
pause/sleep method is the only one that works for my purposes. However, I
totally understand your reluctance to add a sleep to the app. It could
probably work with a smaller interval than 500ms, if that would ease your
concern. If not, I greatly appreciate your time and effort and will
continue to think about other ways to handle the issue.

**Comment by SamInPgh** [comment]
I think I've got this issue taken care of in the plugin, Kurt. I added code
to "grab" the 'playerpref digitalVolumeControl' command by registering an
addDispatch function before it goes to LMS, where I can take prophylactic
action prior to passing it on to its normal destination. Testing is going
very well so far. Thanks again for your help, as always. You can close this
issue.

**Comment by kaaholst** [comment]
Thanks for the info, I'll remove the special handling of the digitalVolumeControl command.

It's probably better this way, otherwise every client would have to implement this behaviour. Even better if it was implemented in LMS (instead of the plugin), so every player would benefit.

**Comment by kaaholst** [comment]
Hi Sam,

Just in case you see this comment.

I may have to take that back (there are no sleeps in the Squeezer source code), For the next version I have implemented an unmute before volume is adjusted via the hardware keys, because I noticed the volume went to zero before being adjusted. This unfortunately didn't work, so I had to implement a delay between the "mixer muting 0" and "mixer volume [adjustment]" commands.

Please let me know it want me to re-enable the Pause playback before toggling "Fixed Volume 100%" functionality.

Incidently I also added a setting for the increments, while doing the new setting for the background volume control.

**Comment by SamInPgh** [comment]
Hi Kurt.

First of all, regarding your change to unmute the player before processing incremental volume adjustments, please take a look at this LMS Issue (now a Discussion) that I opened three months ago:    https://github.com/Logitech/slimserver/discussions/558 

There were some additional back-channel discussions on the subject but, in the end, as you will see, I decided not to pursue the issue further due to Michael's concerns regarding the potential impact on existing plugins or other code that might depend on the current behavior (as un-intuitive as I find it to be).  I instead made changes to the plugin to compensate by doing what you are planning to do --- intercept an incremental volume change while muted and take corrective action by unmuting and restoring the pre-muted volume before processing the change. Your client change should still work with my plugin code, assuming that I first get an unmute, followed by the volume change, and I will be glad to see the same behavior now in all my other Squeezer-controlled players too. However, I still think that the most logical place to change it would have been in the LMS code, where it's literally a one-line change (not counting making it a customizable option, which I feel would not have been necessary). I'm also glad to hear that you are making the change to allow customizing the increment used for the hardware volume keys. I personally prefer a setting of 3 for the clients that support it.

Now, regarding the pause and delay before toggling "Fixed Volume 100%", I have mixed feelings. On the one hand, I liked the way it worked with your test version a little better than what I ended up doing in the plugin.  However, what I have now works the same with all clients (muting the AVR before passing the command to LMS and unmuting it afterwards) and, after all, it is a situation that will rarely be encountered "in the wild". So I think it is best to leave well enough alone at this point. It's good to know, however, that you could be willing to put delays in your code to handle future problems.  ;-)

Thanks for the "heads up". I'll be looking forward to seeing your changes in an upcoming release.

**Comment by kaaholst** [comment]
Hi Sam,

The status of the unmute is that the next version of Squeezer, which is currently in beta, will send an unmute command followed by a delay of 500ms before a volume adjustment is sent. If you are a beta user you should this behaviour also in players not controlled by your plugin. I plan to keep this behaviour in Squeezer for the foreseeable future, because as you say it doesn't conflict with other layers taking the same corrective actions. Even when/if LMS decide to make the change, I will still keep this behaviour in Squeezer, because a substantial number of users will still run an older version of LMS.

Now regarding pause before toggling "Fixed Volume 100%" I will not change anything, until I get a request for it.

And yes, I did see this comment and wanted to reply, but somehow I got away from it. Sorry about that.
You are always welcome to email me directly on kaaholst@gmail.com

Regards
Kurt

**Comment by SamInPgh** [comment]
Sounds good, Kurt. Thanks for the update. I've been running with the latest
beta for awhile now, and your muting change works well with both the Denon
plugin and other players.

Take care,
Sam
